src/wazuh_modules/wm_sca.c

```diff
@@ -3193,3 +3201,59 @@ static int append_msg_to_vm_scat (wm_sca_t * const data, const char * const msg)
     }
     return 0;
 }
+
+/*
+    Thread that sends the integrity hash to the manager every configurable number of seconds.
+    If the manager has a different integrity hash, it will request the agent to send a new
+    sca scan and integrity hash.
+ */
+static void *wm_sca_check_integrity_periodically (wm_sca_t * data) {
+    unsigned int time_sleep = data->integrity_interval;
+
+    while (1) {
+        /* Thread sleeps for 5 minutes until the next time it sends the hash */
+        mdebug2("Sleeping for %d seconds.", (int)time_sleep);
+        wm_delay(1000 * time_sleep);
+
+        /* It doesn't send anything if there are no policies */
+        if (!data->policies) {
+            continue;
+        }
+
+        int i;
+
+        /* Send hash for every policy file */
+        for (i = 0; data->policies[i]; i++) {
+            if (!data->policies[i]->enabled || !data->policies[i]->policy_id) {
+                continue;
+            }
+
+            cJSON *json_integrity = cJSON_CreateObject();
+            int cis_db_index = i;
+
+            cJSON_AddStringToObject(json_integrity, "type", "integrity_check");
+            
+            mdebug1("Calculating hash for scanned results.");
+            char *integrity_hash = wm_sca_hash_integrity(cis_db_index);
```

Comment:
Hello,
After a few checks, it seems there is a race condtion at this point. Long story short: at least two threads , a reader and a writer, are accessing the global variable **cis_db_for_hash**. The solution is just adding mutexes.
> ==7899== Possible data race during read of size 8 at 0x6018060 by thread #10
==7899== Locks held: none
==7899==    at 0x14A0C0: compare_cis_db_info_t_entry (in /var/ossec/bin/wazuh-modulesd)
==7899==    by 0x5C33D20: msort_with_tmp.part.0 (msort.c:83)
==7899==    by 0x5C33A97: msort_with_tmp (msort.c:45)
==7899==    by 0x5C33A97: msort_with_tmp.part.0 (msort.c:54)
==7899==    by 0x5C33FDE: msort_with_tmp (msort.c:45)
==7899==    by 0x5C33FDE: qsort_r (msort.c:297)
==7899==    by 0x14A187: wm_sca_hash_integrity (in /var/ossec/bin/wazuh-modulesd)
==7899==    by 0x14AAA4: wm_sca_check_integrity_periodically (in /var/ossec/bin/wazuh-modulesd)
==7899==    by 0x4C32D06: mythread_wrapper (hg_intercepts.c:389)
==7899==    by 0x59E94A3: start_thread (pthread_create.c:456)
==7899== 
==7899== This conflicts with a previous write of size 1 by thread #3
==7899== Locks held: none
==7899==    at 0x4C35E53: memcpy@GLIBC_2.2.5 (vg_replace_strmem.c:1017)
==7899==    by 0x5C33A81: msort_with_tmp (msort.c:45)
==7899==    by 0x5C33A81: msort_with_tmp.part.0 (msort.c:53)
==7899==    by 0x5C33A81: msort_with_tmp (msort.c:45)
==7899==    by 0x5C33A81: msort_with_tmp.part.0 (msort.c:53)
==7899==    by 0x5C33A97: msort_with_tmp (msort.c:45)
==7899==    by 0x5C33A97: msort_with_tmp.part.0 (msort.c:54)
==7899==    by 0x5C33FDE: msort_with_tmp (msort.c:45)
==7899==    by 0x5C33FDE: qsort_r (msort.c:297)
==7899==    by 0x14A187: wm_sca_hash_integrity (in /var/ossec/bin/wazuh-modulesd)
==7899==    by 0x1513B6: wm_sca_main (in /var/ossec/bin/wazuh-modulesd)
==7899==    by 0x4C32D06: mythread_wrapper (hg_intercepts.c:389)
==7899==  Address 0x6018060 is 432 bytes inside a block of size 800 alloc'd
==7899==    at 0x4C2F1EF: realloc (vg_replace_malloc.c:785)
==7899==    by 0x14F2D0: wm_sca_do_scan (in /var/ossec/bin/wazuh-modulesd)
==7899==    by 0x15137A: wm_sca_main (in /var/ossec/bin/wazuh-modulesd)
==7899==    by 0x4C32D06: mythread_wrapper (hg_intercepts.c:389)
==7899==    by 0x59E94A3: start_thread (pthread_create.c:456)
==7899==  Block was alloc'd by thread #3

Reply:
I have added a mutex for `cis_db_for_hash` in 64874ce, valgrind keeps reporting race conditions, but it appears to be happening because of using the wrapper `w_mutex_lock` for `pthread_mutex_lock`.